src/Voron/Impl/Paging/RvnMemoryMapPager.cs

```diff
@@ -0,0 +1,214 @@
+                    }
+                    metric.IncrementSize(totalUnsynced);
+                }
+            }
+            finally
+            {
+                currentState.Release();
+            }
+        }
+
+        public override void ReleaseAllocationInfo(byte* baseAddress, long size)
+        {
+            base.ReleaseAllocationInfo(baseAddress, size);
+
+            var rc = rvn_unmap(baseAddress, size, DeleteOnClose ? PalDefinitions.True : PalDefinitions.False, out var errorCodeUnmap, out var errorCodeMadvise);
+            // TODO : is madvise don't need - should fail unmap call ? (currently we fail it)
+            if (rc != 0)
+                PalHelper.ThrowLastError(errorCodeUnmap != 0 ? errorCodeUnmap : errorCodeMadvise,
+                    $"Failed to unmap {FileName.FullPath}. rc={rc}. DeleteOnClose={DeleteOnClose}, errorCodeUnmap={errorCodeUnmap}, errorCodeMadvise={errorCodeMadvise}");
+
+            NativeMemory.UnregisterFileMapping(FileName?.FullPath, new IntPtr(baseAddress), size);
+        }
+
+        protected override void DisposeInternal()
+        {
+            var rc = rvn_dispose_handle(FileName.FullPath, _handle, DeleteOnClose ? PalDefinitions.True : PalDefinitions.False, out var errorCodeClose, out var errorCodeUnlink);
+            if (rc == 0) 
+                return;
+            
+            // ignore results.. but warn! we cannot delete the file probably
+            // rc contains either FAIL_UNLINK, FAIL_CLOSE or both of them   
+            if (_logger.IsInfoEnabled)
+                _logger.Info($"Unable to dispose handle for {FileName.FullPath} (ignoring). rc={rc}. DeleteOnClose={DeleteOnClose}, errorCodeClose={errorCodeClose}, errorCodeUnlink={errorCodeUnlink}",
+                    new IOException($"Unable to dispose handle for {FileName.FullPath} (ignoring)."));
+        }
+
+        protected internal override void PrefetchRanges(Win32MemoryMapNativeMethods.WIN32_MEMORY_RANGE_ENTRY* list, int count)
+        {
+            // TODO : Get rid of WIN32_MEMORY_RANGE_ENTRY and use Pal's PrefetchRanges instead
+            // ReSharper disable once UnusedVariable
+            rvn_prefetch_ranges((PalDefinitions.PrefetchRanges*)list, count, out var errorCode);
+            // we explicitly ignore the return code here, this is optimization only
+        }
+
+        protected internal override PagerState AllocateMorePages(long newLength)
+        {
+            if (DisposeOnceRunner.Disposed)
+                ThrowAlreadyDisposedException();
+
+            var rc = rvn_allocate_more_space(newLength, _totalAllocationSize, FileName.FullPath, _handle, _copyOnWriteMode ? MMAP_OPTIONS.CopyOnWrite : MMAP_OPTIONS.None,
+                out var newAddress, out var newLengthAfterAdjustment, out var errorCode);
+
+            if (rc == (int)FAIL_CODES.FAIL_ALLOCATION_NO_RESIZE)
```

Comment:
It is a flag to return null, in order to be consistent with previous implementation (like [here](https://github.com/ravendb/ravendb/blob/v4.1/src/Voron/Impl/Paging/AbstractPager.cs#L315))

Reply:
But this isn't actually an error. This is good, we don't need to do anything because we are big enough.
For that matter, can we move this check to the managed code?

Reply:
if we move this back to managed, we need : _nearest_size_to_page_size.  It will not require pinvoke as we already have the page size from get_system_information.  
Do you want to re-implement _nearest_size_to_page_size in managed ?

Reply:
We already have the page size in managed code, so there is nothing else we need.
We can do that in managed code, yes.
Still need to _verify_ that the desired size is page aligned.